src/lib/TopLevel.hs

```diff
@@ -255,17 +257,115 @@ filterLogs block (Result outs err) = let
   outs' = requiredOuts ++ processLogs (sbLogLevel block) logOuts
   in Result outs' err
 
-summarizeModuleResults :: [Result] -> Result
-summarizeModuleResults results =
-  case [err | Result _ (Failure err) <- results] of
-    [] -> Result allOuts $ Success ()
-    errs -> Result allOuts $ throw ModuleImportErr $ foldMap pprint errs
-  where allOuts = foldMap resultOutputs results
-
-emptyResult :: Result
-emptyResult = Result [] (Success ())
-
-evalFile :: MonadInterblock m => FilePath -> m [(SourceBlock, Result)]
+-- returns a toposorted list of the module's transitive dependencies (including
+-- the module itself) excluding those provided in the set of already known
+-- modules.
+findDepsTransitively
+  :: forall m n. (Topper m, Mut n)
+  => ModuleSourceName -> m n [UModulePartialParse]
+findDepsTransitively initialModuleName = do
+  alreadyLoaded <- M.keysSet . fromLoadedModules <$> withEnv (envLoadedModules . topEnv)
+  flip evalStateT alreadyLoaded $ execWriterT $ go initialModuleName
+  where
+    go :: ModuleSourceName -> WriterT [UModulePartialParse]
+                                (StateT (S.Set ModuleSourceName) (m n)) ()
+    go name = do
+      alreadyVisited <- S.member name <$> get
+      unless alreadyVisited do
+        modify $ S.insert name
+        config <- lift $ lift $ getConfig
+        source <- loadModuleSource config name
+        deps <- lift $ lift $ parseUModuleDepsCached name source
+        mapM_ go deps
+        tell [UModulePartialParse name deps source]
+
+-- What would it look like to abstract away pattern used here and in
+-- `evalPartiallyParsedUModuleCached`? We still want case-by-case control over
+-- keys, eviction policy, etc. Maybe some a type class for caches that implement
+-- query/extend, with `extend` being where the eviction happens?
+parseUModuleDepsCached :: (Mut n, TopBuilder m)
+                       => ModuleSourceName -> File -> m n [ModuleSourceName]
+parseUModuleDepsCached Main file = return $ parseUModuleDeps Main file
+parseUModuleDepsCached name file = do
+  cache <- parsedDeps <$> getCache
+  let req = fHash file
+  case M.lookup name cache of
+    Just (cachedReq, result) | cachedReq == req -> return result
+    _ -> do
+      let result = parseUModuleDeps name file
+      extendCache $ mempty { parsedDeps = M.singleton name (req, result) }
+      return result
+
+evalPartiallyParsedUModuleCached
+  :: (Topper m, Mut n)
+  => UModulePartialParse -> m n (ModuleName n)
+evalPartiallyParsedUModuleCached md@(UModulePartialParse name deps source) = do
+  case name of
+    Main -> evalPartiallyParsedUModule md  -- Don't cache main
+    _ -> do
+      LiftE cache <- withEnv $ LiftE . moduleEvaluations . envCache . topEnv
+      -- TODO: we know that these are sufficient to determine the result of
+      -- module evaluation, but maybe we should actually restrict the
+      -- environment we pass to `evalUModule` so that it can't possibly depend
+      -- on anything else.
```

Comment:
Can you elaborate? I don't think I have sufficient context to understand this comment

Reply:
My point was just that I've made up this cache key based on knowledge of how `evalUModule` works. I know that the result has a functional dependence on `directDeps`. But the function actually has access to much more data, it just chooses not to use it. Maybe in the future it will change so that it does use some of that extra data, and then caching won't be valid. A way around that is to restrict what `evalUModule` has access to so that that can't happen.